src/Kafka/Producer/Types.hs

```diff
@@ -39,3 +40,19 @@ data ProducePartition =
     SpecifiedPartition {-# UNPACK #-} !Int  -- the partition number of the topic
   | UnassignedPartition
   deriving (Show, Eq, Ord, Typeable)
+
+-- | Represents the report of a successfully delivered message.
+data ProducerSuccess = ProducerSuccess
+  { psTopic     :: !TopicName    -- ^ Kafka topic this message was received from
+  , psPartition :: !PartitionId  -- ^ Kafka partition this message was received from
+  , psOffset    :: !Offset       -- ^ Offset within the 'crPartition' Kafka partition
+  , psKey       :: !(Maybe BS.ByteString)
+  , psValue     :: !(Maybe BS.ByteString)
+  }
+  deriving (Eq, Show, Read, Typeable)
+
+-- | Represents the failure to deliver a message.
+data ProducerError = ProducerError
```

Comment:
Yeah sounds good. I rederive all I need from the payload but it's true that other people might not have everything they need in there.

```haskell
DeliverySuccess ProducerRecord Offset
```

Did you mean `ConsumerRecord` here?

Reply:
No, `ProducerRecord`.
I now think that `ProducerRecord` makes more sense because we tell the producer to send  `ProducerRecord`, and it is just logical that it returns that `ProducerRecord` in a delivery report.

If you think about it, using a producer doesn't necessarily mean using a consumer. It would probably be awkward if delivery reports were returning `ConsumerRecord`, do you agree?

Reply:
Hm, I see what you mean, but in the successful case there is more data that one wants. So you have already added `Offset`, but one would also want the `PartitionId` (_not_ the `ProducePartition`).

Otherwise yes I agree let's go with this plan.

Reply:
I see. We have a choice here, either to update `ProducePartition` to `SpecifiedPartition`, or to add `PartitionId` to a constructor like `DeliverySuccess ProducerRecord Offset PartitionId`...

Reply:
I think that returning the partition in `prPartition = SpecifiedPartition x` will be fine because I don't think that at this point we can recover the the original `prPartition` value.